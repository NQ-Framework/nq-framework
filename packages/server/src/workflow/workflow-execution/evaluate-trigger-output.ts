import {
  reducePropertyValuesToObject,
  WorkflowExecutionContext,
  WorkflowExecutionResult,
} from '@nqframework/models';
import { WorkflowTriggerBase } from '@nqframework/models/build/workflow/triggers/workflow-trigger-base';
import { evaluateProperties } from '../../core/utils';

export const evaluateTriggerOutput = async (
  trigger: WorkflowTriggerBase,
  result: WorkflowExecutionResult,
  context: WorkflowExecutionContext,
): Promise<any> => {
  if (!trigger.output || !trigger.output.length) {
    return [];
  }
  const values = await evaluateProperties(trigger.output, { result, context });
  return reducePropertyValuesToObject(values);
};
