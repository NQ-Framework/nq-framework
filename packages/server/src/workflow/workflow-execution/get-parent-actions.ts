import { ActionInstance, ActionLink, Workflow } from '@nqframework/models';

export function getParentActions(
    workflow: Workflow,
    actionInstance: ActionInstance,
): ActionInstance[] {
    const links = workflow.actionLinks as ActionLink[];
    return links
        .filter((al) => al.toName === actionInstance.name)
        .map((link) =>
            workflow.actionInstances!.find((ai) => ai.name === link.fromName),
        ) as ActionInstance[];
}