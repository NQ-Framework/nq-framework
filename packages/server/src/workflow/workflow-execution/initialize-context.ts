import { WorkflowExecutionContext } from '@nqframework/models';

export function initializeContext(
  context: WorkflowExecutionContext,
): WorkflowExecutionContext {
  const workflow = context.workflow;
  const actionInstances = workflow.actionInstances
    ? workflow.actionInstances.filter((ai) => ai.isEnabled)
    : [];

  if (actionInstances.length === 0) {
    throw new Error(
      'Execution context cannot be initialized with zero actions',
    );
  }

  let actionLinks = workflow.actionLinks;
  if (!actionLinks) {
    actionLinks = [];
  } else {
    actionLinks = actionLinks.filter((al) =>
      al.isEnabled &&
      actionInstances.some((ai) => ai.name === al.fromName) &&
      actionInstances.some((ai) => ai.name === al.toName)
    );
  }
  return {
    ...context,
    workflow: { ...workflow, actionInstances, actionLinks },
  };
}
