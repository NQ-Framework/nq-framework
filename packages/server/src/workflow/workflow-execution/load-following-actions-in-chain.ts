import {
  ActionInstance,
  Workflow,
  WorkflowExecutionContext,
} from '@nqframework/models';

export function loadFollowingActionsInChain(
  workflow: Workflow,
  actionInstance: ActionInstance,
  context: WorkflowExecutionContext,
) {
  if (!workflow.actionLinks || workflow.actionLinks.length === 0) {
    return;
  }

  const nextActions = workflow.actionLinks
    .filter((al) => al.fromName === actionInstance.name)
    .map((al) => workflow.actionInstances.find((ai) => ai.name === al.toName));
  nextActions.forEach((nextAction) => {
    if (nextAction !== undefined) {
      context.stack.push(nextAction);
    }
  });
}
