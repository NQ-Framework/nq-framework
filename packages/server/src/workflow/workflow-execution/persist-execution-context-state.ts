import { WorkflowExecutionContext } from '@nqframework/models';
import { getFirebaseApp } from '../../firebase/initialize';

export const persistExecutionContextState = async (
  context: WorkflowExecutionContext,
): Promise<void> => {
  const app = await getFirebaseApp();
  await app
    .firestore()
    .collection('workflowExecutionHistory')
    .doc(context.id)
    .set(context);
};
