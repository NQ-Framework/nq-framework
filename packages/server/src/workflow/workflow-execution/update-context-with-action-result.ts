import {
  ActionInstance,
  ActionResult,
  WorkflowExecutionContext,
} from '@nqframework/models';
import { reducePropertyValuesToObject } from '../../core/utils';

export function updateContextWithActionResult(
  context: WorkflowExecutionContext,
  actionInstance: ActionInstance,
  result: ActionResult,
) {
  context.actions[actionInstance.name as any] = {
    properties: reducePropertyValuesToObject(result.propertyValues),
    values: reducePropertyValuesToObject(result.outputValues),
  };

  context.input = reducePropertyValuesToObject(result.outputValues);
}
