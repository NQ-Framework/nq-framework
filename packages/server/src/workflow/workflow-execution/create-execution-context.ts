import {
  PropertyValue,
  Workflow,
  WorkflowExecutionContext,
} from '@nqframework/models';
import {
  evaluateProperties,
  reducePropertyValuesToObject,
} from '../../core/utils';
import { initializeContext } from './initialize-context';

export async function createExecutionContext(
  input: PropertyValue[],
  workflow: Workflow,
): Promise<WorkflowExecutionContext> {
  let context: WorkflowExecutionContext = {
    isRunning: true,
    startTime: new Date(),
    stack: [],
    actions: {},
    input: {},
    triggerInput: input,
    workflow: workflow,
  };
  const evaluatedProps = await evaluateProperties(input, context);
  context.input = reducePropertyValuesToObject(evaluatedProps);

  context = initializeContext(context);

  return context;
}
