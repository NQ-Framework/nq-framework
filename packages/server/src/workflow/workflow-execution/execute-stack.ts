import {
  WorkflowExecutionContext,
  WorkflowExecutionResult,
} from '@nqframework/models';
import { ActionService } from '../../actions/action.service';
import { loadFollowingActionsInChain } from './load-following-actions-in-chain';
import { updateContextWithActionResult } from './update-context-with-action-result';

export async function executeStack(
  context: WorkflowExecutionContext,
  actionService: ActionService,
): Promise<WorkflowExecutionResult> {
  const workflow = context.workflow;

  while (context.stack.length > 0) {
    const actionInstance = context.stack[context.stack.length - 1];

    const result = await actionService.executeAction(actionInstance, context);

    updateContextWithActionResult(context, actionInstance, result);
    context.stack.pop();

    loadFollowingActionsInChain(workflow, actionInstance, context);
  }

  return {
    data: context.actions,
    finalOutput: context.input,
  };
}
