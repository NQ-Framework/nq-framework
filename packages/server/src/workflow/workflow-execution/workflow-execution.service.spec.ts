import { Test, TestingModule } from '@nestjs/testing';
import { mockWorkflow } from '../mocks/mock-workflow';
import { WorkflowExecutionService } from './workflow-execution.service';
import { createExecutionContext } from './create-execution-context';
import { ActionService } from '../../actions/action.service';
import { getMockExecutionContext } from '../mocks/get-mock-execution-context';
import { executeStack } from './execute-stack';
import { mockExecutionResult } from '../mocks/mock-execution-result';
import { OrganizationService } from '../../organization/organization.service';
import { mockOrganization } from '../mocks/mock-organization';
import { persistExecutionContextState } from "./persist-execution-context-state";


jest.mock('./create-execution-context');
jest.mock('./execute-stack');
jest.mock('./persist-execution-context-state');

describe('WorkflowExecutionService', () => {
  let service: WorkflowExecutionService;
  let organizationService: OrganizationService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        WorkflowExecutionService,
        { provide: ActionService, useValue: { type: 'mockService' } },
        {
          provide: OrganizationService,
          useValue: { getOrganization: jest.fn() },
        },
      ],
    }).compile();

    service = module.get<WorkflowExecutionService>(WorkflowExecutionService);
    organizationService = module.get<OrganizationService>(OrganizationService);
    (persistExecutionContextState as jest.Mock).mockClear();
  });

  it('should call create context', async () => {
    const mock = organizationService.getOrganization as jest.Mock;
    mock.mockImplementation(() => mockOrganization);
    (createExecutionContext as jest.Mock).mockImplementation(() => {
      return getMockExecutionContext();
    });

    (executeStack as jest.Mock).mockImplementation(() => {
      return mockExecutionResult;
    });

    const result = await service.executeWorkflow(
      mockWorkflow,
      [],
      mockWorkflow.triggers[0].id,
      "mock user id"
    );

    expect(createExecutionContext).toHaveBeenCalledWith(
      [],
      mockWorkflow,
      mockOrganization,
      "mock user id"
    );
    expect(executeStack).toHaveBeenCalledWith(
      expect.objectContaining({ isRunning: true }),
      { type: 'mockService' },
    );
    expect(result).toEqual(mockExecutionResult);
    expect(persistExecutionContextState).toHaveBeenCalledTimes(2);
    expect(persistExecutionContextState).toHaveBeenNthCalledWith(1, expect.objectContaining({ isRunning: true }))
    expect(persistExecutionContextState).toHaveBeenLastCalledWith(expect.objectContaining({ isRunning: false }))
  });
  it('should start with items indicated in the trigger', async () => {
    const mockOrg = organizationService.getOrganization as jest.Mock;
    mockOrg.mockImplementation(() => mockOrganization);
    (createExecutionContext as jest.Mock).mockImplementation(() => {
      return getMockExecutionContext();
    });
    (executeStack as jest.Mock).mockImplementation(() => {
      return mockExecutionResult;
    });
    await service.executeWorkflow(
      mockWorkflow,
      [],
      mockWorkflow.triggers[0].id,
      "mock user id"
    );
    expect(createExecutionContext).toHaveBeenCalledWith(
      [],
      mockWorkflow,
      mockOrganization,
      "mock user id"
    );
    expect(executeStack).toHaveBeenCalledWith(
      {
        ...getMockExecutionContext(),
        stack: [mockWorkflow.actionInstances[0]],
      },
      { type: 'mockService' },
    );
    expect(mockOrg).toHaveBeenCalledWith(mockWorkflow.organizationId);
  });

  it('should throw when starting with invalid trigger id', async () => {
    const mockOrg = organizationService.getOrganization as jest.Mock;
    mockOrg.mockImplementation(() => mockOrganization);
    (createExecutionContext as jest.Mock).mockImplementation(() => {
      return getMockExecutionContext();
    });
    await expect(
      service.executeWorkflow(mockWorkflow, [], 'invalid', "mock user id"),
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(persistExecutionContextState).not.toHaveBeenCalled();
  });

  it('should throw when starting with invalid trigger id', async () => {
    const mockOrg = organizationService.getOrganization as jest.Mock;
    mockOrg.mockImplementation(() => mockOrganization);
    const execContext = getMockExecutionContext();
    (createExecutionContext as jest.Mock).mockImplementation(() => {
      return {
        ...execContext,
        workflow: {
          ...execContext.workflow,
          triggers: [
            { ...execContext.workflow.triggers[0], actions: ['invalid'] },
          ],
        },
      };
    });
    await expect(
      service.executeWorkflow(mockWorkflow, [], mockWorkflow.triggers[0].id, "mock user id"),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should throw when starting with invalid organization id', async () => {
    const mockOrg = organizationService.getOrganization as jest.Mock;
    mockOrg.mockImplementation(() => null);
    await expect(
      service.executeWorkflow(mockWorkflow, [], mockWorkflow.triggers[0].id, "mock user id"),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
