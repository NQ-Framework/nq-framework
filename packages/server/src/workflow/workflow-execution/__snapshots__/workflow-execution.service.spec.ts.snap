// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowExecutionService should throw when starting with invalid trigger id 1`] = `"invalid trigger id, cannot start workflow"`;

exports[`WorkflowExecutionService should throw when starting with invalid trigger id 2`] = `"Invalid action name in trigger, cannot start workflow"`;
